### PR TITLE
⚠️  Plugin chaining

### PR DESCRIPTION
**Pull request by Adirio**
Plugin phase 1.5

Implements #1990
Closes: #2036

### CODE REVIEW

File: generate_testdata.sh
Comment (camilamacedo86) [inline generate_testdata.sh:129]: I do not think that we should need to inform  `go/v2` it is the layout f the project already. So, here we would inform just the declarative one.
Also, should we not keep the same name (addon)? have we any doc about it? If yes, we need a top update as well

File: testdata/project-v3-addon/PROJECT
Comment (camilamacedo86) [inline testdata/project-v3-addon/PROJECT:2]: The layout of the project is `go.kubebuilder.io/v3`. The `declarative.kubebuilder.io/v1` is an additional plugin used on top of that and should be tracked in the plugins attr instead: 

```
plugins:
  declarative.kubebuilder.io/v1: {}
```

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
I am ok with that. However, has few points that I do not agree with here:

**1 - Project layout** 

It represents the layout used to scaffold the project. 
Then, I do not agree with:

```
layout: go.kubebuilder.io/v3,declarative.kubebuilder.io/v1
```

The `declarative.kubebuilder.io/v1` plugin is an extra plugin used to do a scaffold on top of the specific subcommand create api of `golang` plugin, the default project layout.

Then in POV, it should be tracked as:

```
plugins:
  declarative.kubebuilder.io/v1: {}
```

Also, I do not think that we should be able/allowed to change/overwrite the default layout. It can be messy and will not allow us to know what was the `base` used to create the project. 

So, the `layout` attr in the config would only allow we have more than one plugin if/when it was used in the init command ( which has the responsibility to create the project base). E.g if we create a project (init) with `go/v2,helm/v1` it would scaffold both bases. And then, for this scenario, we need to describe how we will solve many issues such as: what info will be described in the help? what args will be allowed? I mean, if we scaffold a base `go/v2,helm/v1` will it show the args to create an api valid for the go or helm plugin? Or both? 

**2 - We need to check and update any documentation over the addon as well.** 

c/c @DirectXMan12 @estroz 
/hold

**Comment by Adirio** [comment]
> **1 - Project layout**
> 
> It represents the layout used to scaffold the project.
> Then, I do not agree with:
> 
> ```
> layout: go.kubebuilder.io/v3,declarative.kubebuilder.io/v1
> ```
> 
> The `declarative.kubebuilder.io/v1` plugin is an extra plugin used to do a scaffold on top of the specific subcommand create api of `golang` plugin, the default project layout.
> 
> Then in POV, it should be tracked as:
> 
> ```
> plugins:
>   declarative.kubebuilder.io/v1: {}
> ```
> 
> Also, I do not think that we should be able/allowed to change/overwrite the default layout. It can be messy and will not allow us to know what was the `base` used to create the project.
> 
> So, the `layout` attr in the config would only allow we have more than one plugin if/when it was used in the init command ( which has the responsibility to create the project base). E.g if we create a project (init) with `go/v2,helm/v1` it would scaffold both bases. And then, for this scenario, we need to describe how we will solve many issues such as: what info will be described in the help? what args will be allowed? I mean, if we scaffold a base `go/v2,helm/v1` will it show the args to create an api valid for the go or helm plugin? Or both?

You are mixing this with the choice of providing a different `--plugin` flag per command we talked in the triage meeting. That project is scaffolded as so:
```
kubebuilder init --project-version=3 --plugins="go/v2,declarative" --domain testproject.org --license apache2 --owner "The Kubernetes authors"
kubebuilder create api --group crew --version v1 --kind Captain --controller=true --resource=true --make=false
kubebuilder create api --group crew --version v1 --kind FirstMate --controller=true --resource=true --make=false
kubebuilder create api --group crew --version v1 --kind Admiral --controller=true --resource=true --namespaced=false --make=false
```
And therefore the layout is correct.

> **2 - We need to check and update any documentation over the addon as well.**

There is no documentation regarding that.

File: testdata/project-v3-addon/PROJECT
Comment (Adirio) [reply testdata/project-v3-addon/PROJECT:2]: No, `declarative/v1` is configured for the whole project because there is no way to configure a plkugin for only one command call. That is a future feature tracked in #1941.

File: generate_testdata.sh
Comment (Adirio) [reply generate_testdata.sh:129]: This again is mixing ideas from #1941 and makes no sense at all in the context of this PR.

File: testdata/project-v2-addon/PROJECT
Comment (camilamacedo86) [inline testdata/project-v2-addon/PROJECT:32]: Why it changes the config? Should it really change it?

File: generate_testdata.sh
Comment (camilamacedo86) [inline generate_testdata.sh:109]: I am ok with that. However, has few points that I do not agree with here:

**1 - Project layout** 

It represents the layout used to scaffold the project. 
Then, I do not agree with:

```
layout: go.kubebuilder.io/v3,declarative.kubebuilder.io/v1
```

The `declarative.kubebuilder.io/v1` plugin is an extra plugin used to do a scaffold on top of the specific subcommand create api of `golang` plugin, the default project layout. 

Currently, we can decide what APIs will use or not the addon ($ kubebuilder create api --group crew --version v1 --kind Captain --controller=true --resource=true --pattern=addon) and it is scaffolds are not part of the init/base at all. So, I do not agree that we should call this plugin in the init. 

IMO: We ought to
- we can call the plugin in the create api  which means replace `--pattern=addon` by `--plugins=declarative.kubebuilder.io/v1` for the current command `($ kubebuilder create api --group crew --version v1 --kind Captain --controller=true --resource=true --pattern=addon)`
- then, it should be tracked as:

```
plugins:
  declarative.kubebuilder.io/v1: {}
```

Also, I do not think that we should be able/allowed to change/overwrite the default layout. It can be messy and will not allow us to know what was the `base` used to create the project before we define how we will: 
-  described the help
-  define what args will be allowed

I mean, if we scaffold a base `go/v2,helm/v1` will it show the args to create an api valid for the go or helm plugin? Or both? 

Also, the init command should allow we use only plugins which are `base` ones. I mean, we should not able to init a project with a plugin that is not responsible to scaffold the "core/base" files. 

**2 - We need to check and update any documentation over the addon as well.** 
/hold

Comment (camilamacedo86) [reply generate_testdata.sh:109]: We discussed it a lot today. I will close this comment and add a summary in the review.

**Review by camilamacedo86 — COMMENTED** [review]
@Adirio and I discussed the plugin ecosystem a lot today. Then, a summary of all discussed subjects is tracked in https://github.com/kubernetes-sigs/kubebuilder/issues/2016 which indeed can be helpful for we have a better idea and overview the plugin ecosystem and all.

So, I am ok with we move forward since the current behaviour for the addon would be addressed via  #1941 where we would allow users to choose what API would or not be scaffolded with. 

/hold cancel

File: testdata/project-v2-addon/PROJECT
Comment (Adirio) [reply testdata/project-v2-addon/PROJECT:32]: Project config 2 doesn't have the layout field required for this change. It may not be needed when we allow per-command `--plugin` flags as we will be able to use `kb init --plugins="go/v2" ...` and add the declarative plugin in `kb create api` calls, not requiring the layout field in the project file any more.

File: pkg/plugins/golang/options.go
Comment (camilamacedo86) [inline pkg/plugins/golang/options.go:75]: These changes are breaking changes for the lib. Why do we need to change the options to address this need?

File: pkg/plugins/golang/v3/edit.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/edit.go:48]: It is also a breaking change.

File: pkg/cli/api.go
Comment (camilamacedo86) [inline pkg/cli/api.go:33]: We can overwrite the description currently. Will still it possible?

File: Makefile
Comment (camilamacedo86) [inline Makefile:91]: this pr has many breaking changes; the mojo needs to be updated as the pr title for a more meaningful description. e.g add support to 1..N default plugins via the layout config

File: generate_testdata.sh
Comment (camilamacedo86) [inline generate_testdata.sh:132]: the feature is experimental in kb so I think would be fine.

File: pkg/plugins/internal/machinery/scaffold_test.go
Comment (camilamacedo86) [inline pkg/plugins/internal/machinery/scaffold_test.go:469]: why we are removing the plugins?

File: pkg/plugins/golang/v3/edit.go
Comment (Adirio) [reply pkg/plugins/golang/v3/edit.go:48]: Why is this a breaking change?

File: pkg/plugins/golang/options.go
Comment (Adirio) [reply pkg/plugins/golang/options.go:75]: These flags need to be handled by the cli because every plugin that creates an api will need the GVK, thats why the flags need to be binded by the cli and not by a certain plugin or all plugins will be incompatible as they will all have these flags.

File: pkg/cli/api.go
Comment (Adirio) [reply pkg/cli/api.go:33]: Yes, the `UpdateContext` function was renamed to `UpdateMetadata` (because context was conflicting with Ginkgo) and the signature modified a bit. Instead of having a single struct with input and output fields, now it has one with input fields (currently only the command name) and another one with output fields (currently examples and description).

File: pkg/plugins/internal/machinery/scaffold_test.go
Comment (Adirio) [reply pkg/plugins/internal/machinery/scaffold_test.go:469]: This is part of the old experimental plugin system used for addon.

File: pkg/plugins/golang/v3/edit.go
Comment (camilamacedo86) [reply pkg/plugins/golang/v3/edit.go:48]: It affects the consumer. You are changing the exported methods in many places.

File: Makefile
Comment (camilamacedo86) [reply Makefile:91]: Also, it will break a lot the integration with SDK which is fine, we just need to ensure that all will still compatible. c/c @estroz

**Comment by christopherhein** [comment]
Nice work @Adirio this seems like it will really help to further the componentconfig implementation. We currently have a hacky way to extend the ComponentConfig type that requires a bunch of manual file deletions and removing the kubebuilder marker for crd generation. Currently for a custom config type in KB we have docs - https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/component-config-tutorial/config-type.md#adding-a-new-config-type that explain how to do it manually and this would help to automate that setup and allow custom configs to be easier to develop.

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
The scenario https://github.com/operator-framework/operator-sdk/pull/4581#discussion_r582216189 needs also be solved before we get it merged.

File: pkg/cli/cmd_helpers.go
Comment (camilamacedo86) [inline pkg/cli/cmd_helpers.go:43]: imo we ought to be a warning to allow users that used SDK tool to scaffold the project to still able to use KB tool.

**Comment by Adirio** [comment]
After rebasing when #2037 got merged, `go mod tidy` is bumping the dependencies in the declarative projects when it should not. The only difference from non-declarative projects is relation to modules is the import of `kubernetes-sigs/kubebuilder-declarative-pattern`. Any way to prevent this in combination with `go mod tidy` to bump all [those](https://github.com/kubernetes-sigs/kubebuilder/pull/1991/checks?check_run_id=2008582058) dependencies?

File: pkg/cli/cmd_helpers.go
Comment (Adirio) [reply pkg/cli/cmd_helpers.go:43]: This is considered in the follow up #2019 though the `Bundle` concept. Not being able to resolve a plugin definetely needs to be an error or we will make the users get to broken states.

**Comment by Adirio** [comment]
> After rebasing when #2037 got merged, `go mod tidy` is bumping the dependencies in the declarative projects when it should not. The only difference from non-declarative projects is relation to modules is the import of `kubernetes-sigs/kubebuilder-declarative-pattern`. Any way to prevent this in combination with `go mod tidy` to bump all [those](https://github.com/kubernetes-sigs/kubebuilder/pull/1991/checks?check_run_id=2008582058) dependencies?

I discovered the cause. `go` post-scaffold method was running `go mod tidy` and `declarative` post-scaffold method was running `go get kubernetes-sigs/kubebuilder-declarative-pattern@...`. As the `go mod tidy` was being run before `go get ...` it was importing the latest version of `kubebuilder-declarative-pattern` which bumped a bunch of versions. Later `go get ...` was defining the exact declarative pattern version but the damage was already done. Changing `go get ...` calls to the end of scaffold hooks instead of post-scaffold guarantees that they will all be run before `go mod tidy`.

**Comment by Adirio** [comment]
/close

Closing this in favor of the chain of PRs #2057 - #2058 - #2059 that split this PR into several PRs to split concerns and make reviewer job easier.

**Pull request closed without merging**
